### PR TITLE
fix: Custom transform to strip out unsupported java classes from META-INF/

### LINKED ISSUE #256: Record requires ASM8
_Android Gradle Plugin:_
**7.0.4**

_Gradle:_
**7.0.2**

_Optimizer:_
- [ ]  ProGuard
- [x]  R8

_Have you created the sentry.properties file?_
- [x]  Yes -> root folder of project
- [ ]  No

_SDKs:_
- [x]  sentry-android -> 5.5.2

---
I have the following issue:
I'm finding some incompatibilities when using this plugin with the latest version of moshi. See issue https://github.com/square/moshi/issues/1424#issuecomment-986821726 for details. I get the same error posted in that issue, i.e. `Record requires ASM8`. I can't tell if that's related, however when removing the dependency to the sentry gradle plugin I can compile without any issues.

**Steps to reproduce:**
- Add the following dependencies: 
```
implementation(platform("io.sentry:sentry-bom:5.5.2"))
implementation("io.sentry:sentry-android")
implementation("io.sentry:sentry-android-okhttp")
implementation("io.sentry:sentry-android-timber")
```

- Add plugin `io.sentry.android.gradle` version `3.0.0-beta-3`:
```
id("io.sentry.android.gradle") version "3.0.0-beta.3"
```

- Add configuration
```
sentry {
	includeProguardMapping.set(true)
	includeNativeSources.set(true)
	uploadNativeSymbols.set(true)
	autoUploadProguardMapping.set(true)
	autoUploadNativeSymbols.set(true)
	tracingInstrumentation {
		enabled.set(true)
	}
}
```
- Add moshi codegen and generate some adapters
```
implementation("com.squareup.moshi:moshi-kotlin:1.13.0")
kapt("com.squareup.moshi:moshi-kotlin-codegen:1.13.0")
implementation("com.squareup.retrofit2:converter-moshi:2.9.0")
```

- Run gradle sync and build project

**Actual result:**
- Build error:
```

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeExtDexProductionDebug'.
> Could not resolve all files for configuration ':app:productionDebugRuntimeClasspath'.
   > Failed to transform moshi-1.13.0.jar (com.squareup.moshi:moshi:1.13.0) to match attributes {artifactType=android-dex, asm-transformed-variant=productionDebug, dexing-enable-desugaring=true, dexing-is-debuggable=true, dexing-min-sdk=31, org.gradle.category=library, org.gradle.dependency.bundling=external, org.gradle.jvm.environment=standard-jvm, org.gradle.jvm.version=8, org.gradle.libraryelements=jar, org.gradle.status=release, org.gradle.usage=java-runtime, org.jetbrains.kotlin.platform.type=jvm}.
      > Execution failed for AsmClassesTransform: /Users/filippovigani/.gradle/caches/modules-2/files-2.1/com.squareup.moshi/moshi/1.13.0/da685586facab9eb5c4fb630ce248be14e7da21b/moshi-1.13.0.jar.
         > Record requires ASM8

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
```

**Expected result:**
- Build completes successfully

### PR DESCRIPTION
**Pull request by romtsn**
## :scroll: Description

This PR introduces a custom Artifact Transform, that strips out unsupported java classes on Android (currently latest supported java version in AGP is 11). This is merely a workaround for the issue on AGP itself https://issuetracker.google.com/issues/206655905#comment5 to not wait until v7.2 (where it's fixed) is GA'd.

A picture worth more than a thousand words:
image

## :bulb: Motivation and Context

Closes #256 

## :pencil: Checklist

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [x] No breaking changes

### CODE REVIEW

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/util/Versions.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package io.sentry.android.gradle.util
+
+import com.android.builder.model.Version
+import org.gradle.util.internal.VersionNumber
```
Comment (cortinico) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/util/Versions.kt:4]: Really nit: Not sure we want to depend on `org.gradle.util.internal.` classes.

**Review by cortinico — COMMENTED** [review]
Nice stuff :)

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/transforms/MetaInfStripTransform.kt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                    *
+                    * Since the first two transforms have conflicting attributes, we define a
+                    * disambiguation rule to make Gradle select our transform first.
+                    */
+//                    matchingStrategy.disambiguationRules
+//                        .pickFirst { o1, o2 -> if (o1 > o2) -1 else if (o1 < o2) 1 else 0 }
+                }
+            }
+            // sets meta-inf-stripped attr to false for all .jar artifacts
+            dependencies.artifactTypes.named("jar") {
```
Comment (marandaneto) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/transforms/MetaInfStripTransform.kt:105]: What's about `aar`s?

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/util/Versions.kt (hunk 1 above)
Comment (marandaneto) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/util/Versions.kt:4]: Can we vendor the impl. of VersionNumber within the Plugin?

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/transforms/MetaInfStripTransform.kt (hunk 2 above)
Comment (romtsn) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/transforms/MetaInfStripTransform.kt:105]: the multi-release capability is only applicable for jars I think. I couldn't find any info that you can create a mr-jar within an `aar`, so I guess I'd better leave it like this until someone requests it

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/util/Versions.kt (hunk 1 above)
Comment (cortinico) [reply plugin-build/src/main/kotlin/io/sentry/android/gradle/util/Versions.kt:4]: Yeah I was thinking about something like that or re-implement a simple version comparer

**Comment by romtsn** [comment]
@marandaneto ready for final review, added tests, switched to a custom semver, also fixed the logic along the way

File: plugin-build/src/main/kotlin/io/sentry/android/gradle/transforms/MetaInfStripTransform.kt [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                    val jarEntry = entries.nextElement()
+                    if (jarEntry.name.equals(JarFile.MANIFEST_NAME, ignoreCase = true)) {
+                        // we deal with the manifest as a last step, since we need to know if there
+                        // any multi-release entries remained
+                        continue
+                    }
+
+                    if (jarEntry.name.startsWith(versionsDir, ignoreCase = true)) {
+                        val javaVersion = jarEntry.javaVersion
+                        if (javaVersion > 11) {
```
Comment (marandaneto) [inline plugin-build/src/main/kotlin/io/sentry/android/gradle/transforms/MetaInfStripTransform.kt:68]: Maybe add this to a constant

**Review by marandaneto — APPROVED** [review]
Left a suggestion but LGTM, thanks @romtsn

**Pull request merged**